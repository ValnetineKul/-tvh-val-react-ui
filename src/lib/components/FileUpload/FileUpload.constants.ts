export const DEFAULT_MAX_FILE_SIZE_IN_BYTES = '500 KB';

export type MaxFileSize = `${number}${'KB' | 'MB'}`;
export type FileUploadStatus = 'success' | 'warning' | 'loading';
