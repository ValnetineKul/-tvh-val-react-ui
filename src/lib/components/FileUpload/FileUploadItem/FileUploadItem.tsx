import type { FC } from 'react';
import React from 'react';
import type { DataAttributes } from '../../../types/common';
import IconButton from '../../Buttons/IconButton';
import Icon from '../../Icon';
import { Times } from '../../Icon/icons/functional';
import InlineMessage from '../../InlineMessage';
import Surface from '../../Surfaces/Surface';
import useStyles from './FileUploadItem.styles';

export interface FileUploadItemProps {
  deleteIconProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
  fileName: string;
  fileStatus?: 'success' | 'warning' | 'loading';
  onClick: React.MouseEventHandler<HTMLButtonElement>;
  index?: number;
  readOnly?: boolean;
}

const FileUploadItem: FC<FileUploadItemProps> = ({
  deleteIconProps,
  fileName,
  fileStatus,
  onClick,
  index,
  readOnly,
}) => {
  const { classes } = useStyles();

  return (
    <Surface color="150" className={classes.root} key={index}>
      <InlineMessage className={classes.inlineMessage} status={fileStatus} message={fileName} />
      {!readOnly && (
        <IconButton
          aria-label={`remove-button-${fileName}`}
          icon={<Icon icon={Times} />}
          onClick={onClick}
          {...deleteIconProps}
        />
      )}
    </Surface>
  );
};
export default FileUploadItem;
