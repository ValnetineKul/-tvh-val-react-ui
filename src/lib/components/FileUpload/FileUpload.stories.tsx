import React, { useState } from 'react';
import type { FileUploadStatus } from './FileUpload.constants';
import FileUpload, { getFileId } from './FileUpload';
import type { StoryTemplate } from '../../story-utils';
import { createMeta } from '../../story-utils';

type Props = React.ComponentProps<typeof FileUpload>;

type Args = Omit<Props, 'onChange'>;

const Template = ((args: Args) => {
  const [files, setFiles] = useState<File[] | File | null>(args.value);
  return <FileUpload {...args} value={files} onChange={setFiles} />;
}) as StoryTemplate<Args>;

const commonArgs: Args = {
  value: null,
  label: 'Label',
  fileTypes: ['JPG', 'PNG', 'PDF'],
  maxFileSize: '1 MB',
  variant: 'dragAndDrop',
  additionalRestriction: "Maximum amount of 'lines': 250",
  disabled: false,
  required: false,
};

export const Primitive = Template.bind({});
Primitive.args = {
  ...commonArgs,
  shouldDisplayAllowedTypes: true,
};
Primitive.storyName = 'Single';

export const Forbidden = Template.bind({});
Forbidden.args = {
  ...commonArgs,
  fileTypes: undefined,
  forbiddenFileTypes: ['ZIP'],
  shouldDisplayForbiddenTypes: true,
};
Forbidden.storyName = 'SingleForbiddenFileTypes';

export const Multiple = Template.bind({});
Multiple.args = {
  ...Primitive.args,
  value: [],
  multiple: true,
};

const StatusTemplate = (args: Args) => {
  const [files, setFiles] = useState<File[] | null>([]);
  const [fileStatuses, setFileStatuses] = useState<Record<string, FileUploadStatus>>({});

  const handleChange = (uploadedFiles: File[] | null) => {
    if (!uploadedFiles) return;
    const statuses = uploadedFiles.reduce<Record<string, FileUploadStatus>>((col, file) => {
      col[getFileId(file)] = 'success';
      return col;
    }, {});
    setFiles(uploadedFiles);
    setFileStatuses(statuses);
  };

  return <FileUpload {...args} value={files} onChange={handleChange} fileStatuses={fileStatuses} />;
};

export const MultipleWithStatus = StatusTemplate.bind({});
MultipleWithStatus.args = {
  ...Multiple.args,
};
MultipleWithStatus.storyName = 'MultipleWithStatus';

export default createMeta({
  component: FileUpload,
  title: 'Components/InputFields/FileUpload',
  argTypes: {
    multiple: {
      control: {
        disable: true,
      },
    },
    fileTypes: {
      control: {
        disable: true,
      },
    },
    forbiddenFileTypes: {
      control: {
        disable: true,
      },
    },
    value: {
      control: {
        disable: true,
      },
    },
    variant: {
      options: ['basic', 'dragAndDrop'],
      control: { type: 'radio' },
    },
  },
});
