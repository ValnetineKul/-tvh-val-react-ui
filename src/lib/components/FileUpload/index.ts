import FileUpload from './FileUpload';
import type { FileUploadRef } from './FileUpload';

export default FileUpload;
export type { FileUploadRef };
