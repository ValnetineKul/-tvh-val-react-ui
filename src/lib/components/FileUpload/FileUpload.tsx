import React, { useState, forwardRef, useImperativeHandle, useRef } from 'react';
import { FileUpload as FileUploadIcon } from '../Icon/icons/functional';
import Icon from '../Icon';
import Typography from '../Typography';
import ButtonBase from '../Surfaces/ButtonBase';
import InlineMessage from '../InlineMessage';
import useStyles from './FileUpload.styles';
import FileUploadItem from './FileUploadItem';
import Button from '../Buttons/Button';
import type { DataAttributes } from '../../types/common';
import type { MaxFileSize, FileUploadStatus } from './FileUpload.constants';
import getMaxFileSizeInBytes from './helpers/getMaxFileSizeInBytes';
import getFileExtension from './helpers/getFileExtension';
import { mergeRefs } from '../../utils/refs';

export const getFileId = (file: File) => {
  return `${file.name}/${file.size}`;
};

type Value<Multiple> = Multiple extends true ? File[] : File | null;

export interface FileUploadProps<Multiple extends boolean = false> {
  value: Value<Multiple>;
  fileTypes?: string[];
  forbiddenFileTypes?: string[];
  shouldDisplayAllowedTypes?: boolean;
  shouldDisplayForbiddenTypes?: boolean;
  additionalRestriction?: string;
  fileStatuses?: Record<string, FileUploadStatus>;
  maxFileSize?: MaxFileSize;
  label?: string;
  required?: boolean;
  errorMessage?: string;
  disabled?: boolean;
  className?: string;
  fullWidth?: boolean;
  multiple?: Multiple;
  readOnly?: boolean;
  variant?: 'basic' | 'dragAndDrop';
  fileSizeErrorMessage?: string;
  fileTypeErrorMessage?: string;
  fileDuplicationErrorMessage?: string;
  acceptedFileTypesLabel?: string;
  forbiddenFileTypesLabel?: string;
  maximumFileSizeLabel?: string;
  dragAndDropLabel?: string;
  basicDragAndDropLabel?: string;
  onChange: (files: Value<Multiple>) => void;
  onFilesUpload?: (files: Value<Multiple>) => void;
  onFileDelete?: (file: Value<false>) => void;
  getCustomFileName?: (file: File) => string;
  inputProps?: React.HTMLAttributes<HTMLInputElement> & DataAttributes;
  buttonProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
  deleteIconProps?: React.HTMLAttributes<HTMLButtonElement> & DataAttributes;
  inputRef?: React.Ref<HTMLInputElement>;
}

export interface FileUploadRef {
  hideInternalErrors: () => void;
}

export interface FileUploadComponent {
  <Multiple extends boolean = false>(
    props: FileUploadProps<Multiple> & { ref?: React.Ref<FileUploadRef> }
  ): JSX.Element | null;
}

const FileUploadRaw = <Multiple extends boolean = false>(
  {
    value,
    label,
    fileTypes = [],
    forbiddenFileTypes = [],
    fileStatuses,
    maxFileSize,
    required,
    errorMessage,
    disabled,
    className,
    shouldDisplayAllowedTypes = true,
    shouldDisplayForbiddenTypes = true,
    additionalRestriction,
    fullWidth = false,
    multiple = false as never,
    readOnly = false,
    variant = 'dragAndDrop',
    fileSizeErrorMessage = 'Attachment must be less than',
    fileTypeErrorMessage = 'This file format is not allowed',
    fileDuplicationErrorMessage = 'File with same name has already been uploaded',
    acceptedFileTypesLabel = 'Accepted file types',
    forbiddenFileTypesLabel = 'Not accepted file types',
    maximumFileSizeLabel = 'Maximum file size',
    dragAndDropLabel = 'Drag and drop files or click to upload',
    basicDragAndDropLabel = 'Upload file',
    onChange,
    onFilesUpload,
    onFileDelete,
    getCustomFileName,
    inputProps,
    buttonProps,
    deleteIconProps,
    inputRef,
  }: FileUploadProps<Multiple>,
  ref: React.Ref<FileUploadRef>
): JSX.Element => {
  const { classes, cx } = useStyles({ surface: '150' });
  const [isFileSizeError, setIsFileSizeError] = useState(false);
  const [isFileTypeError, setIsFileTypeError] = useState(false);
  const [isFileDuplicationError, setIsFileDuplicationError] = useState(false);
  const fileInputRef = useRef<HTMLInputElement>(null);

  const handleHideErrors = () => {
    setIsFileSizeError(false);
    setIsFileTypeError(false);
    setIsFileDuplicationError(false);
  };

  useImperativeHandle(ref, () => ({
    hideInternalErrors: handleHideErrors,
  }));

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    handleHideErrors();
    const maxFileSizeInBytes = (() => {
      if (!maxFileSize) return;
      return getMaxFileSizeInBytes(maxFileSize);
    })();

    if (!event.target.files) {
      if (!multiple) {
        onChange(null as Value<Multiple>);
      } else {
        onChange(value);
      }
      return;
    }

    const files = Array.from(event.target.files);

    const allowedExtensions = fileTypes?.map((type) => type.toLowerCase()) || [];
    const forbiddenExtensions = forbiddenFileTypes?.map((type) => type.toLowerCase()) || [];

    const validFiles = files.reduce<File[]>((arr, file) => {
      const extension = getFileExtension(file.name);

      const isAllowedType = () => {
        if (!extension) {
          return false;
        }
        if (fileTypes?.length > 0 && extension) {
          return allowedExtensions.includes(extension);
        }
        return true;
      };

      const isNotForbiddenType = () => {
        if (forbiddenExtensions?.length > 0 && extension) {
          return !forbiddenExtensions.includes(extension);
        }
        return true;
      };

      const isSupportedType = isAllowedType() && isNotForbiddenType();
      const isSupportedSize = maxFileSizeInBytes ? file?.size <= maxFileSizeInBytes : true;
      const isNotDuplicated = multiple
        ? !(value as Value<true>).some((validFile) => validFile.name === file.name)
        : true;

      if (isSupportedType && isSupportedSize && isNotDuplicated) {
        if (getCustomFileName) {
          const newFileName = getCustomFileName(file);
          if (newFileName !== file.name) {
            const renamedFile = new File([file], newFileName, {
              type: file.type,
            });
            return [...arr, renamedFile];
          }
        }
        return [...arr, file];
      }
      setIsFileSizeError(!isSupportedSize);
      setIsFileTypeError(!isSupportedType);
      setIsFileDuplicationError(!isNotDuplicated);

      return arr;
    }, []);

    if (!validFiles.length) {
      return;
    }

    if (!multiple) {
      onChange(validFiles[0] as Value<Multiple>);
      onFilesUpload?.(validFiles[0] as Value<Multiple>);
      return;
    }
    onChange([...(value as Value<true>), ...validFiles] as Value<Multiple>);
    onFilesUpload?.(validFiles as Value<Multiple>);
  };

  const rootClassName = cx(classes.root, className, {
    [classes.disabled]: !!disabled,
    [classes.fullWidth]: fullWidth,
  });

  const errorMessageComposed =
    errorMessage ||
    (isFileSizeError && `${fileSizeErrorMessage} ${maxFileSize}`) ||
    (isFileTypeError && fileTypeErrorMessage) ||
    (isFileDuplicationError && fileDuplicationErrorMessage);

  const mainClassName = cx(classes.main, {
    [classes.hasError]: !!errorMessageComposed,
  });
  const accept = fileTypes.map((type) => `.${type}`).join(', ');

  const renderInput = () => {
    return (
      <input
        ref={mergeRefs([inputRef, fileInputRef])}
        aria-label={label || 'file-upload-input'}
        type="file"
        className={cx(classes.input, {
          [classes.basicInput]: variant === 'basic',
        })}
        disabled={disabled}
        onChange={handleChange}
        accept={accept}
        multiple={multiple}
        tabIndex={variant === 'basic' ? -1 : undefined}
        value=""
        {...inputProps}
      />
    );
  };

  const renderFileInfo = () => {
    return (
      <>
        {shouldDisplayAllowedTypes && fileTypes?.length > 0 && (
          <Typography variant="body400" secondary className={classes.secondaryText}>
            {acceptedFileTypesLabel}: {fileTypes.map((type) => type.toUpperCase()).join(', ')}
          </Typography>
        )}

        {shouldDisplayForbiddenTypes && forbiddenFileTypes?.length > 0 && (
          <Typography variant="body400" secondary className={classes.secondaryText}>
            {forbiddenFileTypesLabel}: {forbiddenFileTypes.map((type) => type.toUpperCase()).join(', ')}
          </Typography>
        )}

        {maxFileSize && (
          <Typography variant="body400" secondary className={classes.secondaryText}>
            {maximumFileSizeLabel}: {maxFileSize}
          </Typography>
        )}

        {additionalRestriction && (
          <Typography variant="body400" secondary className={classes.secondaryText}>
            {additionalRestriction}
          </Typography>
        )}
      </>
    );
  };

  const renderDragAndDrop = () => {
    return (
      <ButtonBase color="150" className={mainClassName} component="div" border disabled={disabled} tabIndex={-1}>
        {renderInput()}

        <Icon className={cx(classes.icon, classes.block)} icon={FileUploadIcon} />
        <Typography className={classes.block}>{dragAndDropLabel}</Typography>
        {renderFileInfo()}
      </ButtonBase>
    );
  };

  const renderBasic = () => {
    return (
      <>
        {renderInput()}
        <div className={classes.basicFileInfo}>{renderFileInfo()}</div>
        <Button
          variant="secondary"
          label={basicDragAndDropLabel}
          startIcon={<Icon icon={FileUploadIcon} />}
          disabled={disabled}
          onClick={() => {
            fileInputRef.current?.click();
          }}
          {...buttonProps}
        />
      </>
    );
  };

  const renderComponent = () => {
    if (variant === 'dragAndDrop') return renderDragAndDrop();
    return renderBasic();
  };

  const isSingle = !multiple && value === null;

  return (
    <div className={rootClassName}>
      {label && (
        <Typography variant="body400" className={classes.label}>
          {label}
          {required && <span className={classes.required}>&nbsp;*</span>}
        </Typography>
      )}

      {!readOnly && (isSingle || multiple) && renderComponent()}

      {errorMessageComposed && (
        <InlineMessage
          className={cx(classes.errorMessage, {
            [classes.basicErrorMessage]: variant === 'basic',
          })}
          message={errorMessageComposed}
          status="error"
          size="sm"
        />
      )}

      {multiple &&
        !disabled &&
        (value as Value<true>).map((file, index) => {
          return (
            <FileUploadItem
              deleteIconProps={deleteIconProps}
              fileName={file.name}
              fileStatus={fileStatuses?.[getFileId(file)]}
              key={index}
              readOnly={readOnly}
              onClick={() => {
                const res = [...(value as Value<true>)];
                res.splice(index, 1);
                onChange(res as Value<Multiple>);
                onFileDelete?.(file);
                handleHideErrors();
              }}
            />
          );
        })}
      {!!value && !multiple && !disabled && !Array.isArray(value) && (
        <FileUploadItem
          fileName={value.name}
          fileStatus={fileStatuses?.[getFileId(value)]}
          readOnly={readOnly}
          onClick={() => {
            onChange(null as Value<Multiple>);
            onFileDelete?.(value);
            handleHideErrors();
          }}
        />
      )}
    </div>
  );
};

const FileUpload: FileUploadComponent = forwardRef(FileUploadRaw);

export default FileUpload;
