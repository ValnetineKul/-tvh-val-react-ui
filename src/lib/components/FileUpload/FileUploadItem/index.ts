import FileUploadItem from './FileUploadItem';

export type { FileUploadItemProps } from './FileUploadItem';
export default FileUploadItem;
