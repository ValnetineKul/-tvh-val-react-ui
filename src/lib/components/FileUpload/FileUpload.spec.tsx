import React from 'react';
import { render, userEvent, screen, waitFor } from '../../test-utils';
import FileUpload, { getFileId } from './FileUpload';
import * as FileUploadItem from './FileUploadItem';

const createFileMock = (name: string, type?: string) => new File([name], name, { type: type || 'image/jpg' });

describe('FileUpload', () => {
  it('Should render correctly', () => {
    const { container } = render(<FileUpload value={null} onChange={jest.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Should render basic variant correctly', () => {
    const { container } = render(<FileUpload value={null} variant="basic" onChange={jest.fn()} />);
    expect(container.firstChild).toMatchSnapshot();
  });
  it('Should show label', () => {
    const { getAllByText } = render(<FileUpload label="Label" value={null} onChange={jest.fn()} />);
    expect(getAllByText('Label')).toHaveLength(1);
  });
  it('Should show label with *', () => {
    const { getAllByText } = render(<FileUpload label="Label" required value={null} onChange={jest.fn()} />);
    expect(getAllByText('*')).toHaveLength(1);
  });
  it('Should show error message', () => {
    const { getByText } = render(<FileUpload errorMessage="Error message" value={null} onChange={jest.fn()} />);
    expect(getByText('Error message')).toBeInTheDocument();
  });
  it('Should hide allowed extensions', () => {
    render(<FileUpload fileTypes={['jpg']} shouldDisplayAllowedTypes={false} value={null} onChange={jest.fn()} />);
    expect(screen.queryByText(/jpg/i)).not.toBeInTheDocument();
  });
  it('Should show additional restriction', () => {
    const { getAllByText } = render(
      <FileUpload label="Label" value={null} onChange={jest.fn()} additionalRestriction="additional" />
    );
    expect(getAllByText('additional')).toHaveLength(1);
  });
  it('Should render only file and hide input for single input', () => {
    const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });
    const { getByText, queryByLabelText } = render(<FileUpload value={file} onChange={jest.fn()} />);

    const input = queryByLabelText('file-upload-input');
    expect(input).not.toBeInTheDocument();
    expect(getByText('file1.txt')).toBeInTheDocument();
  });
  it('Should render only file for readOnly case', () => {
    const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });
    const { getByText, queryByLabelText } = render(<FileUpload value={file} onChange={jest.fn()} readOnly />);

    const input = queryByLabelText('file-upload-input');
    expect(input).not.toBeInTheDocument();
    expect(getByText('file1.txt')).toBeInTheDocument();
  });

  describe('Should trigger onChange', () => {
    it('Single', () => {
      const handleChange = jest.fn();
      const handleUpload = jest.fn();
      const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const { getByLabelText } = render(
        <FileUpload
          onChange={handleChange}
          value={null}
          onFilesUpload={handleUpload}
          getCustomFileName={(f) => `test${f.name}`}
        />
      );
      const input = getByLabelText('file-upload-input') as HTMLInputElement;

      userEvent.upload(input, file);

      expect(input.files).toHaveLength(1);
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenCalledWith(file);

      expect(handleUpload).toHaveBeenCalledTimes(1);
      expect(handleUpload).toHaveBeenCalledWith(file);
    });
    it('Multiple', () => {
      const handleChange = jest.fn();
      const handleUpload = jest.fn();

      const newFile = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const currentFile = new File(['file2'], 'file2.txt', { type: 'text/plain' });
      const { getByLabelText } = render(
        <FileUpload multiple onChange={handleChange} value={[currentFile]} onFilesUpload={handleUpload} />
      );
      const input = getByLabelText('file-upload-input') as HTMLInputElement;

      userEvent.upload(input, newFile);

      expect(input.files).toHaveLength(1);
      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenCalledWith([newFile, currentFile]);

      expect(handleUpload).toHaveBeenCalledTimes(1);
      expect(handleUpload).toHaveBeenCalledWith([newFile]);
    });
  });
  describe('Should show files if has value', () => {
    it('Single', () => {
      const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const { getByText } = render(<FileUpload value={file} onChange={jest.fn()} />);
      expect(getByText('file1.txt')).toBeInTheDocument();
    });
    it('Multiple', () => {
      const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const file2 = new File(['file2'], 'file2.txt', { type: 'text/plain' });
      const { getByText } = render(<FileUpload value={[file1, file2]} multiple onChange={jest.fn()} />);
      expect(getByText('file1.txt')).toBeInTheDocument();
      expect(getByText('file2.txt')).toBeInTheDocument();
    });
  });
  it('Should not show files when disabled', () => {
    const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });
    const { getByLabelText, queryByText } = render(<FileUpload onChange={jest.fn()} value={null} disabled />);
    const input = getByLabelText('file-upload-input') as HTMLInputElement;
    userEvent.upload(input, file1);
    expect(queryByText('file1.txt')).not.toBeInTheDocument();
  });

  describe('Should remove file', () => {
    it('Single', () => {
      const handleChange = jest.fn();
      const handleDelete = jest.fn();
      const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const { getByRole } = render(<FileUpload value={file} onChange={handleChange} onFileDelete={handleDelete} />);

      const removeButton = getByRole('button');
      userEvent.click(removeButton);

      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenCalledWith(null);

      expect(handleDelete).toHaveBeenCalledTimes(1);
      expect(handleDelete).toHaveBeenCalledWith(file);
    });

    it('Multiple', () => {
      const handleChange = jest.fn();
      const handleDelete = jest.fn();
      const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const { getByText } = render(
        <FileUpload value={[file]} multiple onChange={handleChange} onFileDelete={handleDelete} />
      );

      const removeButton = getByText('times.svg');
      userEvent.click(removeButton);

      expect(handleChange).toHaveBeenCalledTimes(1);
      expect(handleChange).toHaveBeenCalledWith([]);

      expect(handleDelete).toHaveBeenCalledTimes(1);
      expect(handleDelete).toHaveBeenCalledWith(file);
    });
  });

  describe('Should prevent uploading if did not pass validation', () => {
    it('Exceed max file size error. Single', async () => {
      const handleChange = jest.fn();
      const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });

      render(<FileUpload value={null} maxFileSize="0.004 KB" onChange={handleChange} />);

      const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

      userEvent.upload(input, file);

      expect(screen.queryByText('file1.txt')).not.toBeInTheDocument();
      await waitFor(() => expect(screen.getByText(/Attachment must be less than/)).toBeInTheDocument());
      expect(handleChange).not.toHaveBeenCalled();
    });
    it('Exceed max file size error. Multiple', async () => {
      const handleChange = jest.fn();
      const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const file2 = new File(['file2'], 'file2.txt', { type: 'text/plain' });

      render(<FileUpload value={[file1]} multiple maxFileSize="0.004 KB" onChange={handleChange} />);

      const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

      userEvent.upload(input, file2);

      expect(screen.queryByText('file2.txt')).not.toBeInTheDocument();
      await waitFor(() => expect(screen.getByText(/Attachment must be less than/)).toBeInTheDocument());
      expect(handleChange).not.toHaveBeenCalled();
    });

    describe('Incorrect extension error', () => {
      it('allowed types', async () => {
        const handleChange = jest.fn();
        const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });

        render(<FileUpload multiple value={[]} onChange={handleChange} fileTypes={['na']} />);

        const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

        userEvent.upload(input, file1);

        expect(screen.queryByText('file1.txt')).not.toBeInTheDocument();
        await waitFor(() => expect(screen.getByText(/This file format is not allowed/)).toBeInTheDocument());
        expect(handleChange).not.toHaveBeenCalled();
      });

      it('forbidden to upload folders (files without extension) for allowed types', async () => {
        const handleChange = jest.fn();
        const file1 = new File(['file1'], 'file1');

        render(<FileUpload value={null} onChange={handleChange} fileTypes={['txt']} />);

        const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

        userEvent.upload(input, file1);

        expect(screen.queryByText('file1')).not.toBeInTheDocument();
        await waitFor(() => expect(screen.getByText(/This file format is not allowed/)).toBeInTheDocument());
        expect(handleChange).not.toHaveBeenCalled();
      });

      it('forbidden types', async () => {
        const handleChange = jest.fn();
        const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });

        render(<FileUpload multiple value={[]} onChange={handleChange} forbiddenFileTypes={['txt']} />);

        const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

        userEvent.upload(input, file1);

        expect(screen.queryByText('file1.txt')).not.toBeInTheDocument();
        await waitFor(() => expect(screen.getByText(/This file format is not allowed/)).toBeInTheDocument());
        expect(handleChange).not.toHaveBeenCalled();
      });

      it('forbidden to upload folders (files without extension) for forbidden types', async () => {
        const handleChange = jest.fn();
        const file1 = new File(['file1'], 'file1');

        render(<FileUpload value={null} onChange={handleChange} forbiddenFileTypes={['txt']} />);

        const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

        userEvent.upload(input, file1);

        expect(screen.queryByText('file1')).not.toBeInTheDocument();
        await waitFor(() => expect(screen.getByText(/This file format is not allowed/)).toBeInTheDocument());
        expect(handleChange).not.toHaveBeenCalled();
      });
    });

    it('Duplication error. Multiple', async () => {
      const handleChange = jest.fn();
      const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const file2 = new File(['file2'], 'file2.txt', { type: 'text/plain' });
      const fileDuplication = new File(['file1'], 'file1.txt', { type: 'text/plain' });

      render(<FileUpload multiple value={[file1, file2]} onChange={handleChange} />);
      const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

      userEvent.upload(input, fileDuplication);

      await waitFor(() =>
        expect(screen.getByText(/File with same name has already been uploaded/)).toBeInTheDocument()
      );
      expect(handleChange).not.toHaveBeenCalled();
    });
  });

  describe('Max file size', () => {
    describe('Size limit message', () => {
      const message = 'Maximum file size: ';

      describe('Single file mode', () => {
        it('Should show size limit message if the restriction is set', () => {
          const maxFileSize = '2 MB';
          const messageWithSize = `${message}${maxFileSize}`;

          render(<FileUpload value={null} maxFileSize={maxFileSize} onChange={jest.fn()} />);

          expect(screen.getByText(messageWithSize)).toBeInTheDocument();
        });

        it('Should hide size limit message if the restriction is not set', () => {
          render(<FileUpload value={null} onChange={jest.fn()} />);

          expect(screen.queryByText(message)).not.toBeInTheDocument();
        });
      });

      describe('Multiple file mode', () => {
        const file = new File(['file1'], 'file1.txt', { type: 'text/plain' });
        const cases: Array<File[][]> = [[[]], [[file]]];

        it.each(cases)('Should show size limit message if the restriction is set for %p', (value) => {
          const maxFileSize = '2 MB';
          const messageWithSize = `${message}${maxFileSize}`;

          render(<FileUpload multiple value={value} maxFileSize={maxFileSize} onChange={jest.fn()} />);

          expect(screen.getByText(messageWithSize)).toBeInTheDocument();
        });

        it.each(cases)('Should hide size limit message if the restriction is not set for %p', (value) => {
          render(<FileUpload multiple value={value} onChange={jest.fn()} />);

          expect(screen.queryByText(message)).not.toBeInTheDocument();
        });
      });
    });

    it('Should render files without input', () => {
      const handleChange = jest.fn();
      const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const file2 = new File(['file2'], 'file2.txt', { type: 'text/plain' });
      const { queryByText, queryByLabelText } = render(
        <FileUpload value={[file1, file2]} multiple onChange={handleChange} readOnly />
      );
      const input = queryByLabelText('file-upload-input');

      expect(input).not.toBeInTheDocument();
      expect(queryByText('file2.txt')).toBeInTheDocument();
      expect(queryByText('file1.txt')).toBeInTheDocument();
      expect(handleChange).not.toHaveBeenCalled();
    });

    it('Should render files with statuses', async () => {
      const handleChange = jest.fn();
      const file1 = new File(['file1'], 'file1.txt', { type: 'text/plain' });
      const fileStatuses = { [getFileId(file1)]: 'success' as const };
      const fileItemSpy = jest.spyOn(FileUploadItem, 'default');
      render(<FileUpload value={[file1]} multiple onChange={handleChange} fileStatuses={fileStatuses} />);

      await waitFor(() =>
        expect(fileItemSpy).toHaveBeenCalledWith(expect.objectContaining({ fileStatus: 'success' }), {})
      );
    });
  });

  describe('Should correctly use getCustomFileName', () => {
    it.each([
      [createFileMock('name.txt'), 'new name', 'new name'],
      [createFileMock('name.txt'), 'name', 'name'],
      [createFileMock('name.txt'), 'name2', 'name2'],
    ])('Should rename optionally %#', async (file: File, newName, expectValue) => {
      const getCustomFileName = jest.fn().mockReturnValue(newName);
      const onChange = jest.fn();

      render(<FileUpload value={null} onChange={onChange} getCustomFileName={getCustomFileName} />);
      const input = screen.getByLabelText('file-upload-input') as HTMLInputElement;

      userEvent.upload(input, file);
      await waitFor(() => expect(onChange.mock.calls[0][0].name).toEqual(expectValue));
    });
  });
});
