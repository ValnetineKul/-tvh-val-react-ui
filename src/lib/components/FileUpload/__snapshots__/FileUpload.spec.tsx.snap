// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileUpload Should render basic variant correctly 1`] = `
<div
  class="tss-8mvvuz-FileUpload-root tss-8-FileUpload-root-ref"
>
  <input
    accept=""
    aria-label="file-upload-input"
    class="tss-1wy2t6d-FileUpload-input-FileUpload-basicInput"
    tabindex="-1"
    type="file"
    value=""
  />
  <div
    class="tss-b8icwg-FileUpload-basicFileInfo"
  />
  <div
    class="tss-1thqjd9-Surface-root-Button-surface"
  >
    <div
      class="tss-pyz8vm-SpinnerWithBackdrop-root"
    >
      <div
        class=""
      >
        <button
          class="MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-disableElevation MuiButtonBase-root tss-1umm6q-Button-root-Button-sizeMd-Button-button mui-10gi7f1-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="tss-1ltka8d-Button-startIcon MuiButton-startIcon MuiButton-iconSizeMedium mui-vbsftn-MuiButton-startIcon"
          >
            <div
              class="tss-w1sjks-Icon-root-Icon-sizeMd"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
                color="inherit"
                focusable="false"
                viewBox="0 0 24 24"
              >
                fileUpload.svg
              </svg>
            </div>
          </span>
          Upload file
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`FileUpload Should render correctly 1`] = `
<div
  class="tss-8mvvuz-FileUpload-root tss-8-FileUpload-root-ref"
>
  <div
    class="MuiButtonBase-root tss-8-FileUpload-main-ref tss-kjklz1-ButtonBase-root-ButtonBase-border-ButtonBase-focusOut-FileUpload-main mui-10d1a0h-MuiButtonBase-root"
    role="button"
    tabindex="-1"
  >
    <input
      accept=""
      aria-label="file-upload-input"
      class="tss-18n9fqr-FileUpload-input"
      type="file"
      value=""
    />
    <div
      class="tss-8-FileUpload-icon-ref tss-d7xcak-Icon-root-Icon-sizeMd-FileUpload-icon-FileUpload-block"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium mui-i4bv87-MuiSvgIcon-root"
        color="inherit"
        focusable="false"
        viewBox="0 0 24 24"
      >
        fileUpload.svg
      </svg>
    </div>
    <p
      class="tss-1-Typography-weightRegular-ref tss-1i0da6j-Typography-root-Typography-weightRegular-FileUpload-block-Typography-body500"
    >
      Drag and drop files or click to upload
    </p>
  </div>
</div>
`;
